### PR TITLE
RavenDB-14086 Fix a race between methods that rely on `LastDatabaseRecordIndex`

### PR DESCRIPTION
**Pull request by karmeli87**

### CODE REVIEW

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 1]
```diff
@@ -3101,6 +3103,33 @@ public async Task WaitForIndexNotification(long index, CancellationToken token)
...
+
+            var task = tcs.Task;
+
+            if (task.IsCompleted)
+                return;
+
+            var result = await Task.WhenAny(task, waitAsync);
+
+            if (result.IsFaulted)
+                throw result.Exception;
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs]: Wrap this in another exception. You are modifying the exception stack trace here.

File: test/SlowTests/Cluster/ClusterIndexNotificationsTest.cs [hunk 2]
```diff
@@ -93,6 +95,22 @@ public async Task ShouldThrowTimeoutException()
...
+            {
+                var indexes = new List<Task>();
+                for (int i = 0; i < 100; i++)
+                {
+                    var index = new Index();
+                    indexes.Add(index.ExecuteAsync(store));
+                }
+
+                await Task.WhenAll(indexes);
+            }
```
Comment (ayende) [inline test/SlowTests/Cluster/ClusterIndexNotificationsTest.cs:114]: Don't you need to verify that they all exist here?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 1 above)
Comment (karmeli87) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs]: Right... maybe `await result` is a cleaner solution?

File: test/SlowTests/Cluster/ClusterIndexNotificationsTest.cs (hunk 2 above)
Comment (karmeli87) [reply test/SlowTests/Cluster/ClusterIndexNotificationsTest.cs:114]: done

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 1 above)
Comment (arekpalinski) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs]: +1 for `await result`

**Review by ayende — APPROVED** [review]

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 3]
```diff
@@ -1219,16 +1242,17 @@ private void NotifyFeaturesAboutValueChange(DatabaseRecord record, long index)
...
 
                     if (taken == false)
                         continue;
                 
-                    LastDatabaseRecordIndex = index;
                     DatabaseShutdown.ThrowIfCancellationRequested();
                     SubscriptionStorage?.HandleDatabaseRecordChange(record);
                     EtlLoader?.HandleDatabaseValueChanged(record);
+
+                    LastDatabaseRecordChangeIndex = index;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DocumentDatabase.cs]: Shouldn't this be `LastValueChangeIndex` ?

Comment (karmeli87) [reply src/Raven.Server/Documents/DocumentDatabase.cs]: You are absolutely correct! I will add a test for this.

**Review by ppekrol — APPROVED** [review]

**Pull request merged**